presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java

```diff
@@ -331,4 +339,32 @@ private Estimate summarizePartitionStatistics(
                             .collect(toImmutableSet()));
         }
     }
+
+    @VisibleForTesting
+    static List<HivePartition> getPartitionsSample(List<HivePartition> partitions, int sampleSize)
+    {
+        checkArgument(sampleSize > 0, "sampleSize is expected to be equal to or greater than zero");
+
+        if (partitions.size() <= sampleSize) {
+            return partitions;
+        }
+
+        Map<String, Integer> hashes = computeHashes(partitions);
+        List<HivePartition> partitionsSortedByHash = new ArrayList<>(partitions);
+        Comparator<HivePartition> comparator = Comparator
+                .comparingInt((HivePartition partition) -> hashes.get(partition.getPartitionId()))
+                .thenComparing(HivePartition::getPartitionId);
+        partitionsSortedByHash.sort(comparator);
+        return ImmutableList.copyOf(partitionsSortedByHash.subList(0, sampleSize));
+    }
+
+    private static Map<String, Integer> computeHashes(List<HivePartition> partitions)
+    {
+        ImmutableMap.Builder<String, Integer> result = ImmutableMap.builder();
+        HashFunction hashFunction = goodFastHash(32);
+        for (HivePartition partition : partitions) {
+            result.put(partition.getPartitionId(), hashFunction.hashString(partition.getPartitionId(), UTF_8).asInt());
```

Comment:
`hashUnencodedChars` is recommended over `hashString(.., encoding)` when cross-language compatibility is not required